// @pulumi/digitalocean/kubernetesCluster.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * ## Import
 *
 * Before importing a Kubernetes cluster, the cluster's default node pool must be tagged with the `terraform:default-node-pool` tag. The provider will automatically add this tag if the cluster has a single node pool. Clusters with more than one node pool, however, will require that you manually add the `terraform:default-node-pool` tag to the node pool that you intend to be the default node pool. Then the Kubernetes cluster and all of its node pools can be imported using the cluster's `id`, e.g.
 *
 * ```sh
 *  $ pulumi import digitalocean:index/kubernetesCluster:KubernetesCluster mycluster 1b8b2100-0e9f-4e8f-ad78-9eb578c2a0af
 * ```
 */
class KubernetesCluster extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["autoUpgrade"] = state ? state.autoUpgrade : undefined;
            inputs["clusterSubnet"] = state ? state.clusterSubnet : undefined;
            inputs["createdAt"] = state ? state.createdAt : undefined;
            inputs["endpoint"] = state ? state.endpoint : undefined;
            inputs["ipv4Address"] = state ? state.ipv4Address : undefined;
            inputs["kubeConfigs"] = state ? state.kubeConfigs : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["nodePool"] = state ? state.nodePool : undefined;
            inputs["region"] = state ? state.region : undefined;
            inputs["serviceSubnet"] = state ? state.serviceSubnet : undefined;
            inputs["status"] = state ? state.status : undefined;
            inputs["surgeUpgrade"] = state ? state.surgeUpgrade : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["updatedAt"] = state ? state.updatedAt : undefined;
            inputs["version"] = state ? state.version : undefined;
            inputs["vpcUuid"] = state ? state.vpcUuid : undefined;
        }
        else {
            const args = argsOrState;
            if (!args || args.nodePool === undefined) {
                throw new Error("Missing required property 'nodePool'");
            }
            if (!args || args.region === undefined) {
                throw new Error("Missing required property 'region'");
            }
            if (!args || args.version === undefined) {
                throw new Error("Missing required property 'version'");
            }
            inputs["autoUpgrade"] = args ? args.autoUpgrade : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["nodePool"] = args ? args.nodePool : undefined;
            inputs["region"] = args ? args.region : undefined;
            inputs["surgeUpgrade"] = args ? args.surgeUpgrade : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["version"] = args ? args.version : undefined;
            inputs["vpcUuid"] = args ? args.vpcUuid : undefined;
            inputs["clusterSubnet"] = undefined /*out*/;
            inputs["createdAt"] = undefined /*out*/;
            inputs["endpoint"] = undefined /*out*/;
            inputs["ipv4Address"] = undefined /*out*/;
            inputs["kubeConfigs"] = undefined /*out*/;
            inputs["serviceSubnet"] = undefined /*out*/;
            inputs["status"] = undefined /*out*/;
            inputs["updatedAt"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(KubernetesCluster.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing KubernetesCluster resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new KubernetesCluster(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of KubernetesCluster.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === KubernetesCluster.__pulumiType;
    }
}
exports.KubernetesCluster = KubernetesCluster;
/** @internal */
KubernetesCluster.__pulumiType = 'digitalocean:index/kubernetesCluster:KubernetesCluster';
