// @pulumi/digitalocean/tag.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * Provides a DigitalOcean Tag resource. A Tag is a label that can be applied to a
 * Droplet resource in order to better organize or facilitate the lookups and
 * actions on it. Tags created with this resource can be referenced in your Droplet
 * configuration via their ID or name.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as digitalocean from "@pulumi/digitalocean";
 *
 * // Create a new tag
 * const foobar = new digitalocean.Tag("foobar", {});
 * // Create a new Droplet in nyc3 with the foobar tag
 * const web = new digitalocean.Droplet("web", {
 *     image: "ubuntu-18-04-x64",
 *     region: "nyc3",
 *     size: "s-1vcpu-1gb",
 *     tags: [foobar.id],
 * });
 * ```
 *
 * ## Import
 *
 * Tags can be imported using the `name`, e.g.
 *
 * ```sh
 *  $ pulumi import digitalocean:index/tag:Tag mytag tagname
 * ```
 */
class Tag extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["databasesCount"] = state ? state.databasesCount : undefined;
            inputs["dropletsCount"] = state ? state.dropletsCount : undefined;
            inputs["imagesCount"] = state ? state.imagesCount : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["totalResourceCount"] = state ? state.totalResourceCount : undefined;
            inputs["volumeSnapshotsCount"] = state ? state.volumeSnapshotsCount : undefined;
            inputs["volumesCount"] = state ? state.volumesCount : undefined;
        }
        else {
            const args = argsOrState;
            inputs["name"] = args ? args.name : undefined;
            inputs["databasesCount"] = undefined /*out*/;
            inputs["dropletsCount"] = undefined /*out*/;
            inputs["imagesCount"] = undefined /*out*/;
            inputs["totalResourceCount"] = undefined /*out*/;
            inputs["volumeSnapshotsCount"] = undefined /*out*/;
            inputs["volumesCount"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Tag.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Tag resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Tag(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Tag.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Tag.__pulumiType;
    }
}
exports.Tag = Tag;
/** @internal */
Tag.__pulumiType = 'digitalocean:index/tag:Tag';
